components/sup/src/manager/mod.rs

```diff
@@ -124,31 +149,73 @@ impl FsCfg {
     }
 }
 
-#[derive(Clone, Default)]
+#[derive(Clone)]
 pub struct ManagerConfig {
     pub auto_update: bool,
+    pub custom_state_path: Option<PathBuf>,
     pub eventsrv_group: Option<ServiceGroup>,
     pub update_url: String,
     pub update_channel: String,
     pub gossip_listen: GossipListenAddr,
+    pub ctl_listen: SocketAddr,
     pub http_listen: http_gateway::ListenAddr,
     pub gossip_peers: Vec<SocketAddr>,
     pub gossip_permanent: bool,
     pub ring: Option<String>,
     pub name: Option<String>,
     pub organization: Option<String>,
     pub watch_peer_file: Option<String>,
+}
+
+impl Default for ManagerConfig {
+    fn default() -> Self {
+        ManagerConfig {
+            auto_update: false,
+            custom_state_path: None,
+            eventsrv_group: None,
+            update_url: "".to_string(),
+            update_channel: "".to_string(),
+            gossip_listen: GossipListenAddr::default(),
+            ctl_listen: ctl_gateway::default_addr(),
+            http_listen: http_gateway::ListenAddr::default(),
+            gossip_peers: vec![],
+            gossip_permanent: false,
+            ring: None,
+            name: None,
+            organization: None,
+            watch_peer_file: None,
+        }
+    }
+}
 
-    custom_state_path: Option<PathBuf>,
+impl ManagerConfig {
+    pub fn sup_root(&self) -> PathBuf {
+        match self.custom_state_path {
+            Some(ref custom) => custom.clone(),
+            None => {
+                match self.name {
+                    Some(ref name) => STATE_PATH_PREFIX.join(name),
+                    None => STATE_PATH_PREFIX.join("default"),
+                }
+            }
+        }
+    }
+}
+
+pub struct ManagerState {
+    /// The configuration used to instantiate this Manager instance
+    pub cfg: ManagerConfig,
+    pub services: Arc<RwLock<Vec<Service>>>,
```

Comment:
Should this be and `Arc` or just an `Rc`?

Reply:
It's an `Arc` for now because it was already an `Arc<RwLock<T>>`. This will get changed to an `Rc` in the next phase of refactoring!